examples/directchat.nim

```diff
@@ -33,56 +31,45 @@ const Help = """
   exit: closes the chat
 """
 
-type
-  CustomData = ref object
-    consoleFd: AsyncFD
-    serveFut: Future[void]
-
-  ChatProto = ref object of LPProtocol
-    customData*: CustomData
-    switch: Switch
-    transp: StreamTransport
-    conn: Connection
-    client: bool
-    connected: bool
-    started: bool
-
-proc id (p: ChatProto): string =
-  if not isNil(p.conn.peerInfo):
-    $p.conn.peerInfo.peerId
+type ChatProto = ref object of LPProtocol
+  switch: Switch          # a single entry point for dialing and listening to peer
+  transp: StreamTransport # transport streams between read & write file descriptor 
+  conn: Connection        # create and close read & write stream
+  connected: bool         # if the node is connected to another peer
+  started: bool           # if the node has started
+
+
+# copied from https://github.com/status-im/nim-beacon-chain/blob/0ed657e953740a92458f23033d47483ffa17ccb0/beacon_chain/eth2_network.nim#L109-L115 
+proc initAddress(T: type MultiAddress, str: string): T =
+  let address = MultiAddress.init(str)
+  if IPFS.match(address) and matchPartial(multiaddress.TCP, address):
+    result = address
   else:
-    "unknown"
+    raise newException(MultiAddressError,
+                         "Invalid bootstrap node multi-address")
+proc dialPeer(p: ChatProto, address: string) {.async.} =
+  let multiAddr = MultiAddress.initAddress(address);
+  let parts = address.split("/")
+  let remotePeer = PeerInfo.init(parts[^1],
+                                 [multiAddr])
+
+  echo &"dialing peer: {multiAddr}"
+  p.conn = await p.switch.dial(remotePeer, ChatCodec)
+  p.connected = true
 
-# forward declaration
-proc readWriteLoop(p: ChatProto) {.async, gcsafe.}
-proc readAndPrint(p: ChatProto) {.async, gcsafe.} =
+proc readAndPrint(p: ChatProto) {.async.} =
   while true:
     while p.connected:
-      # echo &"{p.id} -> "
```

Comment:
The `echo &"{p.id} ->` was supposed to be a prompt, but I couldn't get it to work. This is probably something to add for better UX. At the very least we should add a `TODO` to capture this.